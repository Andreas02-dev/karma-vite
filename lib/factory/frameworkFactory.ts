import type { DiFactory } from '../types/diFactory';
import type { Config } from '../types/karma';
import type { ViteProvider } from './viteServerFactory';

const frameworkFactory: DiFactory<
  [vite: ViteProvider, config: Config],
  ViteProvider
> = (vite, config) => {
  if (config.vite?.autoInit !== false) {
    config.set({
      beforeMiddleware: (config.beforeMiddleware || []).concat(['vite-before']),
      middleware: (config.middleware || []).concat(['vite']),
    });
  }
  config.files = config.files?.map((file) => {
    if (typeof file === 'object' && file.vite) {
      return {
        ...file,
        type: 'module',
        watched: false,
        served: false,
      };
    }
    return file;
  });
  return vite;
};

frameworkFactory.$inject = ['vite', 'config'];

export default frameworkFactory;
