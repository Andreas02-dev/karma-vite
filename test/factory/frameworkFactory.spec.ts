import frameworkFactory from '@/factory/frameworkFactory';
import createInjector from '@test/_utils/createInjector';

describe('frameworkFactory', () => {
  it('should return viteServer', async () => {
    const injector = createInjector();
    const viteServer = await injector.get('vite');
    const framework = await injector.invoke(frameworkFactory);
    expect(framework).toBe(viteServer);
  });

  it('should auto init correctly in default config', async () => {
    const injector = createInjector();
    await injector.invoke(frameworkFactory);
    const config = injector.get('config');
    expect(config).toMatchSnapshot();
  });

  it('should not auto init if vite.autoInit is false in config', async () => {
    const injector = createInjector({
      vite: {
        autoInit: false,
      },
    });
    await injector.invoke(frameworkFactory);
    const config = injector.get('config');
    expect(config).toMatchSnapshot();
  });

  it('rewrite the files property of config correctly', async () => {
    const injector = createInjector({
      files: [
        'a.ts',
        { pattern: 'b.ts', included: false, vite: true },
        { pattern: 'b.ts', served: true, vite: true },
        { pattern: 'c.ts', vite: true },
      ],
    });
    await injector.invoke(frameworkFactory);
    const config = injector.get('config');
    expect(config).toMatchSnapshot();
  });
});
