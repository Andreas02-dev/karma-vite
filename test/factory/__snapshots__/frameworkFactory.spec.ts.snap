// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`frameworkFactory rewrite the files property of config correctly 1`] = `
Config {
  "basePath": "./",
  "beforeMiddleware": Array [
    "vite-before",
  ],
  "files": Array [
    "a.ts",
    Object {
      "included": false,
      "pattern": "b.ts",
      "served": false,
      "type": "module",
      "vite": true,
      "watched": false,
    },
    Object {
      "pattern": "b.ts",
      "served": false,
      "type": "module",
      "vite": true,
      "watched": false,
    },
    Object {
      "pattern": "c.ts",
      "served": false,
      "type": "module",
      "vite": true,
      "watched": false,
    },
  ],
  "frameworks": Array [],
  "middleware": Array [
    "vite",
  ],
  "plugins": Array [],
  "reporters": Array [],
  "urlRoot": "/",
}
`;

exports[`frameworkFactory should auto init correctly in default config 1`] = `
Config {
  "basePath": "./",
  "beforeMiddleware": Array [
    "vite-before",
  ],
  "files": Array [],
  "frameworks": Array [],
  "middleware": Array [
    "vite",
  ],
  "plugins": Array [],
  "reporters": Array [],
  "urlRoot": "/",
}
`;

exports[`frameworkFactory should not auto init if vite.autoInit is false in config 1`] = `
Config {
  "basePath": "./",
  "files": Array [],
  "frameworks": Array [],
  "plugins": Array [],
  "reporters": Array [],
  "urlRoot": "/",
  "vite": Object {
    "autoInit": false,
  },
}
`;
